server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeExecutorService.java

```diff
@@ -230,6 +344,142 @@ private void abortMergeTask(MergeTask mergeTask) {
         }
     }
 
+    class DiskSpaceMonitor implements Runnable {
+        private final RelativeByteSizeValue highStageWatermark;
+        private final ByteSizeValue highStageMaxHeadroom;
+        private final NodeEnvironment.DataPath[] dataPaths;
+
+        DiskSpaceMonitor(
+            RelativeByteSizeValue highStageWatermark,
+            ByteSizeValue highStageMaxHeadroom,
+            NodeEnvironment.DataPath[] dataPaths
+        ) {
+            this.highStageWatermark = highStageWatermark;
+            this.highStageMaxHeadroom = highStageMaxHeadroom;
+            this.dataPaths = dataPaths;
+        }
+
+        @Override
+        public void run() {
+            FsInfo.Path leastAvailablePath = null;
+            IOException fsInfoException = null;
+            if (dataPaths == null) {
+                LOGGER.warn("Cannot read filesystem info because data path is not set in the env");
+                return;
+            }
+            for (NodeEnvironment.DataPath dataPath : dataPaths) {
+                try {
+                    FsInfo.Path fsInfo = getFSInfo(dataPath); // uncached
+                    if (leastAvailablePath == null || leastAvailablePath.getAvailable().getBytes() > fsInfo.getAvailable().getBytes()) {
+                        leastAvailablePath = fsInfo;
+                    }
+                } catch (IOException e) {
+                    if (fsInfoException == null) {
+                        fsInfoException = e;
+                    } else {
+                        fsInfoException.addSuppressed(e);
+                    }
+                }
+            }
+            if (fsInfoException != null) {
+                LOGGER.warn("unexpected exception reading filesystem info", fsInfoException);
+            }
+            if (leastAvailablePath == null) {
+                LOGGER.error("Cannot read filesystem info");
+                return;
```

Comment:
If we cannot read the available disk space (because of some errors), merging is allowed to go on, rather then be stopped.

Reply:
Would we not need to call `updateMaxPriorityLimit(Long.MAX_VALUE)` to get that effect - in case we could read it and then cannot? We can also assert that the limit is not set.

Reply:
Sorry, it was not clear to me, and failed to follow-up for clarifications.

I assumed that it was fine to not do nothing if the fs stats fail to read the available space, assuming that the default (start) disk space budget was Long.MAX_VALUE.

I've pushed https://github.com/elastic/elasticsearch/pull/127613/commits/d07b32b5694a83eb19c5ac3bc26bb584982f886c to explicitly not block any merging in this case instead, and added a test for it.